### ISSUE TITLE
Bug: Incorrect padding error is not properly handled for python3 in security.py (lead to server outage)

### ISSUE THREAD

**Issue by jakeczyz** [onnela-lab/beiwe-backend#171]
Overnight, seemingly bad data from a phone caused problems in my cluster.  It began with Incorrect padding errors thrown by base64 decode called via decode_base64() in security.py.  Here's a stack trace:

```
Error: Incorrect padding
  File "flask/app.py", line 2447, in wsgi_app
    response = self.full_dispatch_request()
  File "flask/app.py", line 1952, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "flask/app.py", line 1821, in handle_user_exception
    reraise(exc_type, exc_value, tb)
  File "flask/_compat.py", line 39, in reraise
    raise value
  File "flask/app.py", line 1950, in full_dispatch_request
    rv = self.dispatch_request()
  File "flask/app.py", line 1936, in dispatch_request
    return self.view_functions[rule.endpoint](**req.view_args)
  File "libs/http_utils.py", line 42, in provide_os_determination_and_call
    return some_function(*args, **kwargs)
  File "libs/user_authentication.py", line 18, in authenticate_and_call
    return some_function(*args, **kwargs)
  File "api/mobile_api.py", line 110, in upload
    uploaded_file = decrypt_device_file(patient_id, uploaded_file, client_private_key, user)
  File "libs/encryption.py", line 131, in decrypt_device_file
    decoded_key = decode_base64(file_data[0])
  File "libs/security.py", line 61, in decode_base64
    return base64.urlsafe_b64decode(data)
  File "python3.6/base64.py", line 133, in urlsafe_b64decode
    return b64decode(s)
  File "python3.6/base64.py", line 87, in b64decode
    return binascii.a2b_base64(s)
```

The function in security.py looks like this:
```
 58 def decode_base64(data: bytes) -> bytes:
 59     """ unpacks url safe base64 encoded string. """
 60     try:
 61         return base64.urlsafe_b64decode(data)
 62     except TypeError as e:
 63         if "Incorrect padding" == str(e):
 64             raise PaddingException()
 65         raise
```

The bug seems to be a simple python2 --> 3 conversion oversight, as py2.7 used to throw a TypeError, whereas 3 now throws a binascii.Error.   So, adding the binascii import and replacing line 62 with something like: `except binascii.Error as e:` should fix this (sorry, I know I should, but I don't have time to do a clone and proper PR...).

Unfortunately, after several hours of this error,  this issue seemed to lead to a single: 
```
AdminShutdown: terminating connection due to administrator command
SSL connection has been closed unexpectedly

  File "django/db/backends/utils.py", line 64, in execute
    return self.cursor.execute(sql, params)
```

Which was followed by all data upload connections to the server failing, apparently due to this error:

```
InterfaceError: connection already closed
  File "django/db/backends/base/base.py", line 231, in _cursor
    return self._prepare_cursor(self.create_cursor(name))
  File "django/db/backends/postgresql/base.py", line 220, in create_cursor
    cursor = self.connection.cursor()
```

Which required me to reboot the EC2 instance.    Not sure if resilience for this was intended/designed, but that might be a separate issue (LMK if I should file it).

**Comment by jakeczyz** [comment]
By the way, not sure if necessarily suggests a Beiwe Android app bug, but the bad data seems to be consistently coming from an Android 10 Samsung SM-N981U1 (Galaxy Note 20 5G).  I'm not sure where to report this possible app bug.

**Comment by jakeczyz** [comment]
Unfortunately, patching security.py as described above lead to a new issue, when the backend apparently tries to store the error in the DB, with Postgres complaining about null bytes.   I.e., this is the new error: 
```
Error: Incorrect padding
  File "libs/security.py", line 61, in decode_base64
    return base64.urlsafe_b64decode(data)
  File "python3.6/base64.py", line 133, in urlsafe_b64decode
    return b64decode(s)
  File "python3.6/base64.py", line 87, in b64decode
    return binascii.a2b_base64(s)

PaddingException: null
  File "libs/encryption.py", line 131, in decrypt_device_file
    decoded_key = decode_base64(file_data[0])
  File "libs/security.py", line 68, in decode_base64
    raise PaddingException()

ValueError: A string literal cannot contain NUL (0x00) characters.
  File "flask/app.py", line 2447, in wsgi_app
    response = self.full_dispatch_request()
  File "flask/app.py", line 1952, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "flask/app.py", line 1821, in handle_user_exception
    reraise(exc_type, exc_value, tb)
  File "flask/_compat.py", line 39, in reraise
    raise value
  File "flask/app.py", line 1950, in full_dispatch_request
    rv = self.dispatch_request()
  File "flask/app.py", line 1936, in dispatch_request
    return self.view_functions[rule.endpoint](**req.view_args)
  File "libs/http_utils.py", line 42, in provide_os_determination_and_call
    return some_function(*args, **kwargs)
  File "libs/user_authentication.py", line 18, in authenticate_and_call
    return some_function(*args, **kwargs)
  File "api/mobile_api.py", line 110, in upload
    uploaded_file = decrypt_device_file(patient_id, uploaded_file, client_private_key, user)
  File "libs/encryption.py", line 133, in decrypt_device_file
    create_decryption_key_error(traceback.format_exc())
  File "libs/encryption.py", line 113, in create_decryption_key_error
    participant=user,
  File "django/db/models/manager.py", line 85, in manager_method
    return getattr(self.get_queryset(), name)(*args, **kwargs)
  File "django/db/models/query.py", line 394, in create
    obj.save(force_insert=True, using=self.db)
  File "database/common_models.py", line 152, in save
    super(AbstractModel, self).save(*args, **kwargs)
  File "django/db/models/base.py", line 808, in save
    force_update=force_update, update_fields=update_fields)
  File "django/db/models/base.py", line 838, in save_base
    updated = self._save_table(raw, cls, force_insert, force_update, using, update_fields)
  File "django/db/models/base.py", line 924, in _save_table
    result = self._do_insert(cls._base_manager, using, fields, update_pk, raw)
  File "django/db/models/base.py", line 963, in _do_insert
    using=using, raw=raw)
  File "django/db/models/manager.py", line 85, in manager_method
    return getattr(self.get_queryset(), name)(*args, **kwargs)
  File "django/db/models/query.py", line 1079, in _insert
    return query.get_compiler(using=using).execute_sql(return_id)
  File "django/db/models/sql/compiler.py", line 1112, in execute_sql
    cursor.execute(sql, params)
  File "django/db/backends/utils.py", line 64, in execute
    return self.cursor.execute(sql, params)
```

**Comment by aware-ravi-bhankharia** [comment]
+1 - also facing this issue. I ended up adding 'except binascii.Error: raise PaddingException()' and ignoring exceptions from the db write.

**Comment by biblicabeebli** [comment]
(hoo boy you have hit the jackpot complex error here...)

Zerothly: thank you two for posting about this; I know this class of error, it is super hard to debug, and that followup detail with the null characters is actually helpful in diagnosing what is going on.

Firstly: we just pushed a huge update containing many bug fixes to master.  No promises, but please check if anything is resolved by deploying this update and check back in, I would super appreciate that.

_Just a reminder: any data being decrypted should be treated as personally identifiable information and should not be posted onto the forum. If it is to be transmitted at all it should be done so carefully.  I would recommend Firefox Send but that is... down.  hm.  Please get in touch out-of-band, eli@zagaran.com, if this needs to happen._

I'm happy to put in the time diagnosing this particular issue in detail because it is in an absolutely critical code path.

### Some Context:
Each line of data is an encrypted binary blob wrapped in a base64 encoding, [you can learn about base64 padding here](https://en.wikipedia.org/wiki/Base64#Output_padding).

In `database/profiling_models.py` there are 3 database tables (Django models) that interact with these and are supposed to store errors for later debugging:

- `EncryptionErrorMetadata`
- `LineEncryptionError`
- `DecryptionKeyError`

`LineEncryptionError` is special in that it is purely for debugging on non-production servers and is guarded against by using a special runtime variable, `IS_STAGING`.

### Okay, so what's the problem here?

1) The error here results in a 500-class error (any unhandled python error will do this) and is therefore likely to repeat when the app attempts to upload the file again.  The apps only retire and delete files when it receives a 200 code from the server.  Fortunately, the app will proceed to the next file rather than crash or loop endlessly on one file, and data is also updated in a random order to improve uploaded data coverage if the device is somehow constrained in its upload bandwidth.

2) There are several known cases where we return a 200 code to the app so that it _stops_ uploading a file.  A non exhaustive list of those cases is: known junk files generated by various versions of android, empty files, files that only have decryption keys but no data, files with bad or missing decryption keys, and **any errors that occur in those same locations but present as base64 padding errors.**  A good example of why is that a missing decryption key, regardless of reason, is unrecoverable.

3) Why does this happen at all?  The apps can be killed (or crash, or encounter io problems) at any moment in time.  The app does its best, but shit happens.  We know of no bugs within reasonable app runtime constraints that cause this issue.  (There is a known scenario involving a user-identifying key file getting corrupted that we have seen occasionally, but that is necessarily an out-of-scope systemic error.)

4) Wait, not all errors corrupt a whole file, it might just be a single line!  We handle this and preserve uncorrupted lines.  The `LineEncryptionError` database table is for storing these errors for analysis later.

### ... so what do I do now?

1) Update the backend, see if anything changes, let me know if it disappears.  I'm looking at the changes in the encryption.py file, I don't think there is much change there.  There were changes to base64 encoding code in libs/security.py that may cover your issue.

2) we may have a new bug!  There is no scenario *at all* that makes the existence of null characters in a padding check acceptable.  The apps never use null characters.

3) That database connection error.  It is possible there is an interaction with using Django's database layer on its own.  We have updated the version of Django used in the codebase (substantially) so this may be resolved as well.

Please let me know if the issue resolves, or if it doesn't.  I'm watching this issue.

**Comment by jakeczyz** [comment]
Thanks for the detailed explanation of what's going on, Eli.  I updated to latest master and will monitor, circling back to let you know if any errors surface.

**Comment by jakeczyz** [comment]
Hi Eli, unfortunately, after the update, Sentry is reporting a new error.  Any advice? 

```
UndefinedColumn: column database_chunkregistry.deleted does not exist
LINE 1: SELECT "database_chunkregistry"."id", "database_chunkregistr...
                                              ^

  File "django/db/backends/utils.py", line 64, in execute
    return self.cursor.execute(sql, params)
ProgrammingError: column database_chunkregistry.deleted does not exist
LINE 1: SELECT "database_chunkregistry"."id", "database_chunkregistr...
                                              ^

  File "libs/file_processing.py", line 212, in upload_binified_data
    chunk = ChunkRegistry.objects.get(chunk_path=chunk_path)
  File "django/db/models/manager.py", line 85, in manager_method
    return getattr(self.get_queryset(), name)(*args, **kwargs)
  File "django/db/models/query.py", line 374, in get
    num = len(clone)
  File "django/db/models/query.py", line 232, in __len__
    self._fetch_all()
  File "django/db/models/query.py", line 1121, in _fetch_all
    self._result_cache = list(self._iterable_class(self))
  File "django/db/models/query.py", line 53, in __iter__
    results = compiler.execute_sql(chunked_fetch=self.chunked_fetch)
  File "django/db/models/sql/compiler.py", line 899, in execute_sql
    raise original_exception
  File "django/db/models/sql/compiler.py", line 889, in execute_sql
    cursor.execute(sql, params)
  File "django/db/backends/utils.py", line 64, in execute
    return self.cursor.execute(sql, params)
  File "django/db/utils.py", line 94, in __exit__
    six.reraise(dj_exc_type, dj_exc_value, traceback)
  File "django/utils/six.py", line 685, in reraise
    raise value.with_traceback(tb)
  File "django/db/backends/utils.py", line 64, in execute
    return self.cursor.execute(sql, params)
```

**Comment by biblicabeebli** [comment]
Did you terminate and deploy a new data processing manager/worker server(s)?

**Comment by jakeczyz** [comment]
All I did was pull latest master, apply my elasticbeanstalk config diff, and `eb deploy`.  What other steps should I take to update?

**Comment by biblicabeebli** [comment]
The launch script (hm, maybe its time to rename it to something like
"cluster manager") has several commands, including terminate and deploy
worker and manager servers.  Any time you update the EB servers you need to
terminate and redeploy your data processing servers.  Should only take a
couple minutes, termination is essentially instantaneous.

(There are exceptions to the rule, but those require a high-level
understanding of the changes at play, so you should just do it every time.)

The bug you encountered is a database schema mismatch between what your
processing servers expect and the schema updates when you deployed to eb
servers.  New processing server will solve that, if not ... then it is a
bug and I will fix it.

**Comment by jakeczyz** [comment]
Thanks. Recreating the server seems to have squashed that error.  I'm now seeing, less frequently, this one: 
```FirebaseNotCredentialed: Firebase is not configured, cannot queue notifications.
  File "services/celery_push_notifications.py", line 70, in create_push_notification_tasks
    raise FirebaseNotCredentialed("Firebase is not configured, cannot queue notifications.")
```

**Comment by biblicabeebli** [comment]
ooooooops that's my bad, redesigning that thing's spammyness got
overwhelmed in the clutter of other forecast (push notifications) code.
Though, that code was also just got reengjneered on friday sooo...

**Comment by jakeczyz** [comment]
No worries!  Though, as it's using up our Sentry quota, please do let me know if there's an update to master coming soon or mitigation you recommend.  Thanks!

**Comment by biblicabeebli** [comment]
Hold on, that doesn't make sense.  If you deployed from master (which was just updated with an unrelated set of fixes that you can push out if you want) then you should not be getting that error at all.  That error should only occur enabled if you pushed from the current state of the firebase-config branch.

**Comment by biblicabeebli** [comment]
..... Or you can just deploy the new bugfix I just put on master....

**Comment by biblicabeebli** [comment]
Update:
There may have been multiple different errors occurring.  on #177 I believe I encountered a new padding exception that I eventually identified as a very misleading "padding exception".  I tested some error cases in a newer version of python (3.8, the base64 library is part of the python standard library), and it emits a much better error message.

I'm not 100% sure what the best behavior for that case is, the "fix" is to detect the case specially and raise a different kind of error.

This isn't finalized or tested yet, it is on the bug-fixes branch.

Let me know if anything changes once this makes it onto master.  I'll post here when that happens.
